File: sdk-typescript/packages/transcribe/src/transcribe.ts
/**
 * WAVE SDK - Transcribe API
 *
 * Audio and video transcription with speaker diarization.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
} from '@wave-av/core';

export * from './transcribe-types';
import type {
  CreateTranscriptionRequest,
  ListTranscriptionsParams,
  Speaker,
  TranscriptExportFormat,
  Transcription,
  TranscriptionModel,
  TranscriptionSegment,
  UpdateTranscriptionRequest,
} from './transcribe-types';
export type { TranscriptionStatus, TranscriptionModel, Transcription, TranscriptionSegment, TranscriptionWord, Speaker, CreateTranscriptionRequest, UpdateTranscriptionRequest, ListTranscriptionsParams, TranscriptExportFormat } from './transcribe-types';

export class TranscribeAPI {
  private readonly client: WaveClient;
  private readonly basePath = '/v1/transcribe';

  constructor(client: WaveClient) {
    this.client = client;
  }

  /**
   * Create a transcription job
   *
   * Requires: transcribe:create permission
   */
  async create(request: CreateTranscriptionRequest): Promise<Transcription> {
    return this.client.post<Transcription>(this.basePath, request);
  }

  /**
   * Get a transcription by ID
   *
   * Requires: transcribe:read permission
   */
  async get(transcriptionId: string): Promise<Transcription> {
    return this.client.get<Transcription>(`${this.basePath}/${transcriptionId}`);
  }

  /**
   * Update a transcription
   *
   * Requires: transcribe:update permission
   */
  async update(
    transcriptionId: string,
    request: UpdateTranscriptionRequest
  ): Promise<Transcription> {
    return this.client.patch<Transcription>(
      `${this.basePath}/${transcriptionId}`,
      request
    );
  }

  /**
   * Remove a transcription
   *
   * Requires: transcribe:remove permission (server-side RBAC enforced)
   */
  async remove(transcriptionId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${transcriptionId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * List transcriptions
   *
   * Requires: transcribe:read permission
   */
  async list(
    params?: ListTranscriptionsParams
  ): Promise<PaginatedResponse<Transcription>> {
    return this.client.get<PaginatedResponse<Transcription>>(this.basePath, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }

  /**
   * Get transcription segments
   *
   * Requires: transcribe:read permission
   */
  async getSegments(
    transcriptionId: string,
    params?: PaginationParams & {
      start_time?: number;
      end_time?: number;
      speaker_id?: number;
    }
  ): Promise<PaginatedResponse<TranscriptionSegment>> {
    return this.client.get<PaginatedResponse<TranscriptionSegment>>(
      `${this.basePath}/${transcriptionId}/segments`,
      { params: params as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Update a segment
   *
   * Requires: transcribe:update permission
   */
  async updateSegment(
    transcriptionId: string,
    segmentId: string,
    updates: { text?: string; speaker?: string; speaker_id?: number }
  ): Promise<TranscriptionSegment> {
    return this.client.patch<TranscriptionSegment>(
      `${this.basePath}/${transcriptionId}/segments/${segmentId}`,
      updates
    );
  }

  /**
   * Merge segments
   *
   * Requires: transcribe:update permission
   */
  async mergeSegments(
    transcriptionId: string,
    segmentIds: string[]
  ): Promise<TranscriptionSegment> {
    return this.client.post<TranscriptionSegment>(
      `${this.basePath}/${transcriptionId}/segments/merge`,
      { segment_ids: segmentIds }
    );
  }

  /**
   * Split a segment
   *
   * Requires: transcribe:update permission
   */
  async splitSegment(
    transcriptionId: string,
    segmentId: string,
    splitTime: number
  ): Promise<{ first: TranscriptionSegment; second: TranscriptionSegment }> {
    return this.client.post(
      `${this.basePath}/${transcriptionId}/segments/${segmentId}/split`,
      { split_time: splitTime }
    );
  }

  /**
   * Get speakers
   *
   * Requires: transcribe:read permission
   */
  async getSpeakers(transcriptionId: string): Promise<Speaker[]> {
    return this.client.get<Speaker[]>(
      `${this.basePath}/${transcriptionId}/speakers`
    );
  }

  /**
   * Update speaker label
   *
   * Requires: transcribe:update permission
   */
  async updateSpeaker(
    transcriptionId: string,
    speakerId: number,
    label: string
  ): Promise<Speaker> {
    return this.client.patch<Speaker>(
      `${this.basePath}/${transcriptionId}/speakers/${speakerId}`,
      { label }
    );
  }

  /**
   * Merge speakers
   *
   * Requires: transcribe:update permission
   */
  async mergeSpeakers(
    transcriptionId: string,
    speakerIds: number[],
    newLabel?: string
  ): Promise<Speaker> {
    return this.client.post<Speaker>(
      `${this.basePath}/${transcriptionId}/speakers/merge`,
      { speaker_ids: speakerIds, label: newLabel }
    );
  }

  /**
   * Export transcription
   *
   * Requires: transcribe:read permission
   */
  async exportTranscription(
    transcriptionId: string,
    format: TranscriptExportFormat,
    options?: {
      include_timestamps?: boolean;
      include_speakers?: boolean;
      paragraph_breaks?: boolean;
    }
  ): Promise<{ url: string; expires_at: string }> {
    return this.client.post(`${this.basePath}/${transcriptionId}/export`, {
      format,
      ...options,
    });
  }

  /**
   * Get plain text transcript
   *
   * Requires: transcribe:read permission
   */
  async getText(
    transcriptionId: string,
    options?: { include_speakers?: boolean; paragraph_breaks?: boolean }
  ): Promise<string> {
    const result = await this.client.get<{ text: string }>(
      `${this.basePath}/${transcriptionId}/text`,
      { params: options }
    );
    return result.text;
  }

  /**
   * Search within a transcription
   *
   * Requires: transcribe:read permission
   */
  async search(
    transcriptionId: string,
    query: string,
    options?: { case_sensitive?: boolean; whole_word?: boolean }
  ): Promise<
    Array<{
      segment_id: string;
      text: string;
      start_time: number;
      end_time: number;
      highlight_ranges: Array<{ start: number; end: number }>;
    }>
  > {
    return this.client.post(`${this.basePath}/${transcriptionId}/search`, {
      query,
      ...options,
    });
  }

  /**
   * Start real-time transcription
   *
   * Requires: transcribe:realtime permission
   */
  async startRealtime(
    streamId: string,
    options?: {
      language?: string;
      model?: TranscriptionModel;
      speaker_diarization?: boolean;
    }
  ): Promise<{
    session_id: string;
    websocket_url: string;
    expires_at: string;
  }> {
    return this.client.post(`${this.basePath}/realtime/start`, {
      stream_id: streamId,
      ...options,
    });
  }

  /**
   * Stop real-time transcription
   *
   * Requires: transcribe:realtime permission
   */
  async stopRealtime(sessionId: string): Promise<Transcription> {
    return this.client.post<Transcription>(
      `${this.basePath}/realtime/${sessionId}/stop`
    );
  }

  /**
   * Get real-time session status
   *
   * Requires: transcribe:read permission
   */
  async getRealtimeStatus(
    sessionId: string
  ): Promise<{
    status: 'active' | 'paused' | 'stopped';
    duration: number;
    word_count: number;
    segments_count: number;
  }> {
    return this.client.get(`${this.basePath}/realtime/${sessionId}`);
  }

  /**
   * Wait for transcription to complete
   */
  async waitForReady(
    transcriptionId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (transcription: Transcription) => void;
    }
  ): Promise<Transcription> {
    const pollInterval = options?.pollInterval || 2000;
    const timeout = options?.timeout || 1800000; // 30 minutes
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const transcription = await this.get(transcriptionId);

      if (options?.onProgress) {
        options.onProgress(transcription);
      }

      if (transcription.status === 'ready') {
        return transcription;
      }

      if (transcription.status === 'failed') {
        throw new Error(
          `Transcription failed: ${transcription.error || 'Unknown error'}`
        );
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Transcription timed out after ${timeout}ms`);
  }

  /**
   * Detect language from audio
   *
   * Requires: transcribe:read permission
   */
  async detectLanguage(
    sourceUrl: string
  ): Promise<{
    detected_language: string;
    confidence: number;
    alternatives: Array<{ language: string; confidence: number }>;
  }> {
    return this.client.post(`${this.basePath}/detect-language`, {
      source_url: sourceUrl,
    });
  }

  /**
   * Get supported languages
   *
   * Requires: transcribe:read permission
   */
  async getSupportedLanguages(): Promise<
    Array<{
      code: string;
      name: string;
      native_name: string;
      models: TranscriptionModel[];
    }>
  > {
    return this.client.get(`${this.basePath}/languages`);
  }

  /**
   * Estimate transcription cost
   *
   * Requires: transcribe:read permission
   */
  async estimateCost(
    durationSeconds: number,
    model: TranscriptionModel = 'standard',
    options?: { speaker_diarization?: boolean }
  ): Promise<{
    estimated_cost: number;
    currency: string;
    breakdown: Record<string, number>;
  }> {
    return this.client.post(`${this.basePath}/estimate`, {
      duration_seconds: durationSeconds,
      model,
      ...options,
    });
  }
}

/**
 * Create a Transcribe API instance
 */
export function createTranscribeAPI(client: WaveClient): TranscribeAPI {
  return new TranscribeAPI(client);
}

File: sdk-python/wave_sdk/transcribe.py
"""WAVE SDK - Transcribe API. Audio/video transcription with speaker diarization."""
from __future__ import annotations
import time
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Speaker(BaseModel):
    id: str; label: str; segments_count: int = 0; total_duration: float = 0

class TranscriptionSegment(BaseModel):
    id: str; start_time: float; end_time: float; text: str; speaker_id: str | None = None; confidence: float = 1.0

class Transcription(BaseModel):
    id: str; asset_id: str | None = None; source_url: str | None = None; status: str; language: str; model: str = "default"; segments: list[TranscriptionSegment] | None = None; speakers: list[Speaker] | None = None; duration_seconds: float | None = None; created_at: str; updated_at: str

class TranscribeAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/transcribe"
    def create(self, source_url: str | None = None, asset_id: str | None = None, language: str = "en", model: str = "default", **kwargs: Any) -> Transcription: return Transcription(**self._client.post(self._base, json={"source_url": source_url, "asset_id": asset_id, "language": language, "model": model, **kwargs}))
    def get(self, transcription_id: str) -> Transcription: return Transcription(**self._client.get(f"{self._base}/{transcription_id}"))
    def list(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def update(self, transcription_id: str, **kwargs: Any) -> Transcription: return Transcription(**self._client.patch(f"{self._base}/{transcription_id}", json=kwargs))
    def remove(self, transcription_id: str) -> None: self._client.delete(f"{self._base}/{transcription_id}")
    def get_segments(self, transcription_id: str) -> list[TranscriptionSegment]: return [TranscriptionSegment(**s) for s in self._client.get(f"{self._base}/{transcription_id}/segments")]
    def update_segment(self, transcription_id: str, segment_id: str, **kwargs: Any) -> TranscriptionSegment: return TranscriptionSegment(**self._client.patch(f"{self._base}/{transcription_id}/segments/{segment_id}", json=kwargs))
    def split_segment(self, transcription_id: str, segment_id: str, split_time: float) -> list[TranscriptionSegment]: return [TranscriptionSegment(**s) for s in self._client.post(f"{self._base}/{transcription_id}/segments/{segment_id}/split", json={"split_time": split_time})]
    def merge_segments(self, transcription_id: str, segment_ids: list[str]) -> TranscriptionSegment: return TranscriptionSegment(**self._client.post(f"{self._base}/{transcription_id}/segments/merge", json={"segment_ids": segment_ids}))
    def get_speakers(self, transcription_id: str) -> list[Speaker]: return [Speaker(**s) for s in self._client.get(f"{self._base}/{transcription_id}/speakers")]
    def update_speaker(self, transcription_id: str, speaker_id: str, **kwargs: Any) -> Speaker: return Speaker(**self._client.patch(f"{self._base}/{transcription_id}/speakers/{speaker_id}", json=kwargs))
    def merge_speakers(self, transcription_id: str, speaker_ids: list[str]) -> Speaker: return Speaker(**self._client.post(f"{self._base}/{transcription_id}/speakers/merge", json={"speaker_ids": speaker_ids}))
    def get_text(self, transcription_id: str, include_speakers: bool = False, include_timestamps: bool = False) -> dict: return self._client.get(f"{self._base}/{transcription_id}/text", params={"include_speakers": include_speakers, "include_timestamps": include_timestamps})
    def search(self, transcription_id: str, query: str) -> list[dict]: return self._client.get(f"{self._base}/{transcription_id}/search", params={"q": query})
    def export_transcription(self, transcription_id: str, format: str = "srt") -> dict: return self._client.get(f"{self._base}/{transcription_id}/export", params={"format": format})
    def detect_language(self, source_url: str) -> dict: return self._client.post(f"{self._base}/detect-language", json={"source_url": source_url})
    def estimate_cost(self, duration_seconds: float, model: str = "default") -> dict: return self._client.post(f"{self._base}/estimate", json={"duration_seconds": duration_seconds, "model": model})
    def get_supported_languages(self) -> list[dict]: return self._client.get(f"{self._base}/languages")
    def start_realtime(self, stream_id: str, language: str = "en", **kwargs: Any) -> dict: return self._client.post(f"{self._base}/realtime/start", json={"stream_id": stream_id, "language": language, **kwargs})
    def stop_realtime(self, stream_id: str) -> dict: return self._client.post(f"{self._base}/realtime/stop", json={"stream_id": stream_id})
    def get_realtime_status(self, stream_id: str) -> dict: return self._client.get(f"{self._base}/realtime/{stream_id}")
    def wait_for_ready(self, transcription_id: str, poll_interval: float = 2.0, timeout: float = 600.0) -> Transcription:
        start = time.time()
        while time.time() - start < timeout:
            t = self.get(transcription_id)
            if t.status == "ready": return t
            if t.status == "failed": raise RuntimeError("Transcription failed")
            time.sleep(poll_interval)
        raise TimeoutError(f"Transcription timed out after {timeout}s")
